packages/fx-core/src/plugins/resource/cicd/plugin.ts

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import { Context } from "@microsoft/teamsfx-api/build/v2";
+import { Inputs, v2, Result, FxError, Platform } from "@microsoft/teamsfx-api";
+import { FxCICDPluginResultFactory as ResultFactory } from "./result";
+import { CICDProviderFactory } from "./providers/factory";
+import { ProviderKind } from "./providers/enums";
+import { questionNames } from "./questions";
+import { InternalError, NoProjectOpenedError } from "./errors";
+import { generateBuildScript } from "./utils/buildScripts";
+import { telemetryHelper } from "./utils/telemetry-helper";
+
+export class CICDImpl {
+  public async addCICDWorkflows(
+    context: Context,
+    inputs: Inputs,
+    envInfo: v2.EnvInfoV2
+  ): Promise<Result<any, FxError>> {
+    // 1. Key inputs (envName, provider, template) x (hostingType, ).
+    if (!inputs.projectPath) {
+      throw new NoProjectOpenedError();
+    }
+    const projectPath = inputs.projectPath;
+    // By default(VSC), get env name from plugin's own `target-env` question.
+    let envName = inputs[questionNames.Environment];
+    if (inputs.platform == Platform.CLI) {
+      // In CLI, get env name from the default `env` question.
+      envName = envInfo.envName;
+    }
+    const providerName = inputs[questionNames.Provider];
+    const templateNames = inputs[questionNames.Template] as string[];
+    if (!envName || !providerName || templateNames.length == 0) {
+      throw new InternalError("Some preconditions of inputs are not met.");
```

Comment:
Better to indicate what precondition is not met.

Reply:
The error is kind of for robustness consideration, i.e., the error should never be throwed out since it's just about inner contract. So, I prefer to leave it as it is.